In [338]:
import pandas as pd
import numpy as np

### Reading data 📖

In [339]:
sleep_data=pd.read_csv('/Users/irasafonik/Desktop/kaggle/Sleep_health_and_lifestyle_dataset.csv')
sleep_data.tail()

Person ID  Gender  Age Occupation  Sleep Duration  Quality of Sleep  \
369        370  Female   59      Nurse             8.1                 9   
370        371  Female   59      Nurse             8.0                 9   
371        372  Female   59      Nurse             8.1                 9   
372        373  Female   59      Nurse             8.1                 9   
373        374  Female   59      Nurse             8.1                 9   

     Physical Activity Level  Stress Level BMI Category Blood Pressure  \
369                       75             3   Overweight         140/95   
370                       75             3   Overweight         140/95   
371                       75             3   Overweight         140/95   
372                       75             3   Overweight         140/95   
373                       75             3   Overweight         140/95   

     Heart Rate  Daily Steps Sleep Disorder  
369          68         7000    Sleep Apnea  
370          68         7000    Sleep Apnea  
371          68         7000    Sleep Apnea  
372          68         7000    Sleep Apnea  
373          68         7000    Sleep Apnea

### Statistical information 🔍

In [340]:
#head() for display the first 5 rows  
sleep_data.head().style.set_properties(**{'background-color': '#E6E6FA',
                                          'color': '#171719'}) #for colored output

In [341]:
formatted_shape = sleep_data.shape

print('The dimention of data is:', formatted_shape)

The dimention of data is: (374, 13)


##### Observations 📋

##### 🟡 There are 374 rows , 13 coulmns

In [342]:
sleep_data.info() # for empty and type of values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374 entries, 0 to 373
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Person ID                374 non-null    int64  
 1   Gender                   374 non-null    object 
 2   Age                      374 non-null    int64  
 3   Occupation               374 non-null    object 
 4   Sleep Duration           374 non-null    float64
 5   Quality of Sleep         374 non-null    int64  
 6   Physical Activity Level  374 non-null    int64  
 7   Stress Level             374 non-null    int64  
 8   BMI Category             374 non-null    object 
 9   Blood Pressure           374 non-null    object 
 10  Heart Rate               374 non-null    int64  
 11  Daily Steps              374 non-null    int64  
 12  Sleep Disorder           155 non-null    object 
dtypes: float64(1), int64(7), object(5)
memory usage: 38.1+ KB


##### Observations 📋

##### 🟡 There are 5 columns are string the rest are numeric in terms of datatype.

In [343]:
#for statistical info
sleep_data.describe().style.background_gradient(cmap='Purples') #for colored output

In [344]:
#for statistical info including string values
sleep_data.describe(include='O').style.set_properties(**{'background-color': '#E6E6FA',
                                                      'color': '#171719'})  

### Exploratory Data Analysis **(EDA)** 📊

In [345]:
columns_name = sleep_data.columns 
print('Columns name:', columns_name) #for show names of columns

Columns name: Index(['Person ID', 'Gender', 'Age', 'Occupation', 'Sleep Duration',
       'Quality of Sleep', 'Physical Activity Level', 'Stress Level',
       'BMI Category', 'Blood Pressure', 'Heart Rate', 'Daily Steps',
       'Sleep Disorder'],
      dtype='object')


In [346]:
sleep_data.duplicated().sum()

0

In [347]:
sleep_data.drop_duplicates() #in case of any dublicates

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  
0           126/83          77         4200            NaN  
1           125/80          75        10000            NaN  
2           125/80          75        10000            NaN  
3           140/90          85         3000    Sleep Apnea  
4           140/90          85         3000    Sleep Apnea  
..             ...         ...          ...            ...  
369         140/95          68         7000    Sleep Apnea  
370         140/95          68         7000    Sleep Apnea  
371         140/95          68         7000    Sleep Apnea  
372         140/95          68         7000    Sleep Apnea  
373         140/95          68         7000    Sleep Apnea  

[374 rows x 13 columns]

In [348]:
sleep_data.isna().sum()

Person ID                    0
Gender                       0
Age                          0
Occupation                   0
Sleep Duration               0
Quality of Sleep             0
Physical Activity Level      0
Stress Level                 0
BMI Category                 0
Blood Pressure               0
Heart Rate                   0
Daily Steps                  0
Sleep Disorder             219
dtype: int64

#### Observations 📋

##### 🟡 There is 219 null values

In [349]:
# Fill missing values in the 'Sleep Disorder' column with 'None'
sleep_data['Sleep Disorder'] = sleep_data['Sleep Disorder'].fillna('None')
sleep_data.head(5)

Person ID Gender  Age            Occupation  Sleep Duration  \
0          1   Male   27     Software Engineer             6.1   
1          2   Male   28                Doctor             6.2   
2          3   Male   28                Doctor             6.2   
3          4   Male   28  Sales Representative             5.9   
4          5   Male   28  Sales Representative             5.9   

   Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                 6                       42             6   Overweight   
1                 6                       60             8       Normal   
2                 6                       60             8       Normal   
3                 4                       30             8        Obese   
4                 4                       30             8        Obese   

  Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  
0         126/83          77         4200           None  
1         125/80          75        10000           None  
2         125/80          75        10000           None  
3         140/90          85         3000    Sleep Apnea  
4         140/90          85         3000    Sleep Apnea

In [350]:
sleep_data.nunique() #unique values for each column

Person ID                  374
Gender                       2
Age                         31
Occupation                  11
Sleep Duration              27
Quality of Sleep             6
Physical Activity Level     16
Stress Level                 6
BMI Category                 4
Blood Pressure              25
Heart Rate                  19
Daily Steps                 20
Sleep Disorder               3
dtype: int64

## Deep dive into each column

### Gender column

In [351]:
gender=sleep_data['Gender'].unique()
print('The unique values of Sex column are :', gender)

The unique values of Sex column are : ['Male' 'Female']


In [352]:
gender_count=sleep_data['Gender'].value_counts().reset_index()
gender_count

Gender  count
0    Male    189
1  Female    185

### Age column

In [353]:
sleep_data['Age'].describe()

count    374.000000
mean      42.184492
std        8.673133
min       27.000000
25%       35.250000
50%       43.000000
75%       50.000000
max       59.000000
Name: Age, dtype: float64

In [354]:
age_count=sleep_data['Age'].value_counts().reset_index()
age_count.head(5)

Age  count
0   43     34
1   44     30
2   37     20
3   38     20
4   50     20

In [355]:
age_count.tail(5)

Age  count
26   48      3
27   55      2
28   56      2
29   34      2
30   27      1

In [356]:
pd.set_option('display.max_columns', None)

gender_age = sleep_data.groupby(['Gender', 'Age']).size().unstack(fill_value=0)
gender_age

Age     27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  \
Gender                                                                       
Female   0   0   2   2   1   0   2   2   1  10  15   9   3   2   0   2   3   
Male     1   5  11  11  17  17  11   0  11   2   5  11  12   2  12   7  31   

Age     44  45  48  49  50  51  52  53  54  55  56  57  58  59  
Gender                                                          
Female  11  14   1   9  20   8   9  17   7   2   2   9   6  16  
Male    19   0   2   2   0   0   0   0   0   0   0   0   0   0

#### Insights💡
#### 🔸 Among females: the highest number of individuals are aged from 50 to 53 years, with the peak at 50 years (20 cases).
#### 🔸 Among males: the highest number of individuals are aged from 29 to 33 years and from 39 to 44 years, with the peak at 43 years (31 cases).

### Occupation column

In [357]:
print('The unique values of Occupation column are', sleep_data['Occupation'].unique())

The unique values of Occupation column are ['Software Engineer' 'Doctor' 'Sales Representative' 'Teacher' 'Nurse'
 'Engineer' 'Accountant' 'Scientist' 'Lawyer' 'Salesperson' 'Manager']


In [358]:
Occupation_count=sleep_data['Occupation'].value_counts().reset_index()
Occupation_count

Occupation  count
0                  Nurse     73
1                 Doctor     71
2               Engineer     63
3                 Lawyer     47
4                Teacher     40
5             Accountant     37
6            Salesperson     32
7      Software Engineer      4
8              Scientist      4
9   Sales Representative      2
10               Manager      1

### Sleep Duration column

In [359]:
sleep_data['Sleep Duration'].describe()

count    374.000000
mean       7.132086
std        0.795657
min        5.800000
25%        6.400000
50%        7.200000
75%        7.800000
max        8.500000
Name: Sleep Duration, dtype: float64

In [360]:
Sleep_Duration_count=sleep_data['Sleep Duration'].value_counts().reset_index()
Sleep_Duration_count.head(5)

Sleep Duration  count
0             7.2     36
1             6.0     31
2             7.8     28
3             6.5     26
4             6.1     25

In [361]:
gender_sleep = sleep_data.groupby(['Gender','Sleep Duration']).size().unstack(fill_value=0)
gender_sleep

Sleep Duration  5.8  5.9  6.0  6.1  6.2  6.3  6.4  6.5  6.6  6.7  6.8  6.9  \
Gender                                                                       
Female            2    1   13   15    6    0    2   12   18    3    4    2   
Male              0    3   18   10    6   13    7   14    2    2    1    1   

Sleep Duration  7.1  7.2  7.3  7.4  7.5  7.6  7.7  7.8  7.9  8.0  8.1  8.2  \
Gender                                                                       
Female           13   22    0    1    1    0    0    0    1   13   13   11   
Male              6   14   14    4    4   10   24   28    6    0    2    0   

Sleep Duration  8.3  8.4  8.5  
Gender                         
Female            5   14   13  
Male              0    0    0

#### Insights💡
#### 🔸 Among females: the highest value is 7.2 hours (22 cases).
#### 🔸 Among males: the highest value is 7.8 hours (28 cases).

### Quality of sleep column

In [362]:
print('The unique values of Quality of Sleep column are', sleep_data['Quality of Sleep'].unique())

The unique values of Quality of Sleep column are [6 4 7 5 8 9]


In [363]:
Quality_of_Sleep_count=sleep_data['Quality of Sleep'].value_counts().reset_index()
Quality_of_Sleep_count

Quality of Sleep  count
0                 8    109
1                 6    105
2                 7     77
3                 9     71
4                 5      7
5                 4      5

### Physical Activity Level column

In [364]:
sleep_data['Physical Activity Level'].describe()

count    374.000000
mean      59.171123
std       20.830804
min       30.000000
25%       45.000000
50%       60.000000
75%       75.000000
max       90.000000
Name: Physical Activity Level, dtype: float64

In [365]:
Physical_Activity_Level=sleep_data['Physical Activity Level'].value_counts().reset_index()
Physical_Activity_Level

Physical Activity Level  count
0                        60     70
1                        30     68
2                        45     68
3                        75     67
4                        90     67
5                        40      6
6                        55      6
7                        35      4
8                        50      4
9                        70      3
10                       42      2
11                       32      2
12                       80      2
13                       65      2
14                       85      2
15                       47      1

### Stress Level column

In [366]:
print('The unique values of Stress Level column are', sleep_data['Stress Level'].unique())

The unique values of Stress Level column are [6 8 7 4 3 5]


In [367]:
Stress_Level_counts=sleep_data['Stress Level'].value_counts().reset_index()
Stress_Level_counts

Stress Level  count
0             3     71
1             8     70
2             4     70
3             5     67
4             7     50
5             6     46

In [368]:
position_stress = sleep_data.groupby(['Occupation', 'Stress Level']).size().unstack(fill_value=0)
position_stress

Stress Level           3   4   5   6   7   8
Occupation                                  
Accountant             2  26   0   3   6   0
Doctor                 4   0   2  32   0  33
Engineer              34   9  16   1   3   0
Lawyer                 0   0  44   3   0   0
Manager                0   0   1   0   0   0
Nurse                 31   4   0   2   4  32
Sales Representative   0   0   0   0   0   2
Salesperson            0   0   0   0  32   0
Scientist              0   0   0   2   0   2
Software Engineer      0   0   2   1   0   1
Teacher                0  31   2   2   5   0

#### Insights💡
#### 🔸Doctors: Report stress levels predominantly at 8 and 6, with 33 and 32 instances, respectively.
#### 🔸Engineers: Mainly report stress levels of 3 and 5, with 34 and 16 instances, respectively.
#### 🔸Teachers: Predominantly report stress levels at 4, with 31 instances.

### BMI Category column

In [369]:
print('The unique values of BMI Category are', sleep_data['BMI Category'].unique())

The unique values of BMI Category are ['Overweight' 'Normal' 'Obese' 'Normal Weight']


In [370]:
# Replace 'Normal' category in 'BMI Category' column with 'Normal Weight' for clarity
sleep_data['BMI Category'] = sleep_data['BMI Category'].replace({'Normal': 'Normal Weight'})

# Count the occurrences of each BMI category and reset index to create a DataFrame
BMI_Category_count = sleep_data['BMI Category'].value_counts().reset_index()
BMI_Category_count

BMI Category  count
0  Normal Weight    216
1     Overweight    148
2          Obese     10

In [371]:
position_bmi=sleep_data.groupby('Occupation')['BMI Category'].value_counts().reset_index()
position_bmi

Occupation   BMI Category  count
0             Accountant  Normal Weight     31
1             Accountant     Overweight      6
2                 Doctor  Normal Weight     67
3                 Doctor          Obese      4
4               Engineer  Normal Weight     60
5               Engineer     Overweight      3
6                 Lawyer  Normal Weight     43
7                 Lawyer          Obese      2
8                 Lawyer     Overweight      2
9                Manager     Overweight      1
10                 Nurse     Overweight     66
11                 Nurse  Normal Weight      7
12  Sales Representative          Obese      2
13           Salesperson     Overweight     32
14             Scientist     Overweight      4
15     Software Engineer  Normal Weight      2
16     Software Engineer          Obese      1
17     Software Engineer     Overweight      1
18               Teacher     Overweight     33
19               Teacher  Normal Weight      6
20               Teacher          Obese      1

#### Insights💡
#### 🔸Nurses: Exhibit the highest count of overweight individuals, with 66 instances.
#### 🔸Doctors: Predominantly fall within the normal weight category, with 67 instances.
#### 🔸Engineers: predominantly belong to the normal weight category, with 60 instances.
#### 🔸Teachers: predominantly fall into the overweight category, with 33 instances

### Blood Pressure column

In [372]:
print('The unique values of Blood Pressure column are', sleep_data['Blood Pressure'].unique())

The unique values of Blood Pressure column are ['126/83' '125/80' '140/90' '120/80' '132/87' '130/86' '117/76' '118/76'
 '128/85' '131/86' '128/84' '115/75' '135/88' '129/84' '130/85' '115/78'
 '119/77' '121/79' '125/82' '135/90' '122/80' '142/92' '140/95' '139/91'
 '118/75']


In [373]:
Blood_Pressure_count=sleep_data['Blood Pressure'].value_counts().reset_index()
Blood_Pressure_count

Blood Pressure  count
0          130/85     99
1          140/95     65
2          125/80     65
3          120/80     45
4          115/75     32
5          135/90     27
6          140/90      4
7          125/82      4
8          132/87      3
9          128/85      3
10         126/83      2
11         115/78      2
12         139/91      2
13         142/92      2
14         119/77      2
15         135/88      2
16         129/84      2
17         128/84      2
18         131/86      2
19         117/76      2
20         130/86      2
21         118/75      2
22         121/79      1
23         122/80      1
24         118/76      1

#### Note 📋
#### 🔴 Ideal blood pressure systolic (upper number) : less than 120 , diastolic (bottom number) : less than 80

#### 🔴 Normal systolic (upper number) : in range (120 - 129) , diastolic (bottom number) : in range (80 - 84)

#### 🔴 Otherwise, blood pressure is high

In [374]:
# Group the data by 'Occupation', 'Gender', and 'Blood Pressure', count occurrences, and reset the index
position_blood = sleep_data.groupby(['Occupation', 'Gender', 'Blood Pressure']).size().reset_index(name='count')
position_blood

Occupation  Gender Blood Pressure  count
0             Accountant  Female         115/75     26
1             Accountant  Female         115/78      2
2             Accountant  Female         119/77      2
3             Accountant  Female         130/85      6
4             Accountant    Male         118/76      1
5                 Doctor  Female         118/75      2
6                 Doctor    Male         120/80     38
7                 Doctor    Male         125/80     27
8                 Doctor    Male         139/91      2
9                 Doctor    Male         142/92      2
10              Engineer  Female         125/80     32
11              Engineer    Male         117/76      1
12              Engineer    Male         120/80      4
13              Engineer    Male         121/79      1
14              Engineer    Male         122/80      1
15              Engineer    Male         125/80      6
16              Engineer    Male         130/85     17
17              Engineer    Male         132/87      1
18                Lawyer  Female         128/85      1
19                Lawyer  Female         135/88      1
20                Lawyer    Male         125/82      2
21                Lawyer    Male         130/85     42
22                Lawyer    Male         135/88      1
23               Manager  Female         125/82      1
24                 Nurse  Female         117/76      1
25                 Nurse  Female         120/80      1
26                 Nurse  Female         125/82      1
27                 Nurse  Female         126/83      1
28                 Nurse  Female         130/86      2
29                 Nurse  Female         132/87      2
30                 Nurse  Female         140/95     65
31  Sales Representative    Male         140/90      2
32           Salesperson    Male         130/85     32
33             Scientist  Female         128/85      2
34             Scientist  Female         131/86      2
35     Software Engineer    Male         120/80      2
36     Software Engineer    Male         126/83      1
37     Software Engineer    Male         140/90      1
38               Teacher  Female         115/75      6
39               Teacher  Female         130/85      2
40               Teacher  Female         135/90     27
41               Teacher    Male         128/84      2
42               Teacher    Male         129/84      2
43               Teacher    Male         140/90      1

#### Insights💡
#### 🔸 Nurses (Female): Have the highest count of elevated blood pressure readings, with 65 instances recorded at 140/95.
#### 🔸 Lawyers (Male): Show a significant number of high blood pressure readings, with 42 instances recorded at various levels, including 130/85 and 135/88.
#### 🔸 Engineer (Female): Most common blood pressure reading is 125/80, with 32 instances.

### Heart Rate column

In [375]:
Heart_Rate_count=sleep_data['Heart Rate'].value_counts().reset_index()
Heart_Rate_count

Heart Rate  count
0           68     94
1           70     76
2           72     69
3           65     67
4           75     36
5           78      5
6           85      3
7           80      3
8           84      2
9           83      2
10          73      2
11          67      2
12          74      2
13          77      2
14          81      2
15          76      2
16          69      2
17          86      2
18          82      1

### Daily Steps column

In [376]:
sleep_data['Daily Steps'].describe()

count      374.000000
mean      6816.844920
std       1617.915679
min       3000.000000
25%       5600.000000
50%       7000.000000
75%       8000.000000
max      10000.000000
Name: Daily Steps, dtype: float64

In [377]:
Daily_Steps_count=sleep_data['Daily Steps'].value_counts().reset_index()
Daily_Steps_count

Daily Steps  count
0          8000    101
1          6000     68
2          5000     68
3          7000     66
4         10000     36
5          5500      4
6          3000      3
7          3500      3
8          4000      3
9          6800      3
10         4800      2
11         7300      2
12         7500      2
13         4200      2
14         3300      2
15         5600      2
16         5200      2
17         4100      2
18         3700      2
19         6200      1

In [378]:
# Group the data by 'Gender', and 'Daily Steps', count occurrences, and reset the index
gender_steps = sleep_data.groupby(['Gender', 'Daily Steps']).size().unstack(fill_value=0)
gender_steps

Daily Steps  3000   3300   3500   3700   4000   4100   4200   4800   5000   \
Gender                                                                       
Female           0      1      0      0      2      2      1      0     36   
Male             3      1      3      2      1      0      1      2     32   

Daily Steps  5200   5500   5600   6000   6200   6800   7000   7300   7500   \
Gender                                                                       
Female           2      4      0     34      0      1     63      2      2   
Male             0      0      2     34      1      2      3      0      0   

Daily Steps  8000   10000  
Gender                     
Female           1     34  
Male           100      2

#### Insights💡
#### 🔸 The most common daily step counts for females are 7000 and 10000, with 63 and 34 instances
#### 🔸 Male individuals primarily achieve daily step counts of 8000, with 100 instances, followed by 6000 steps, with 34 instances

### Sleep Disorder column

In [379]:
print('The unique values of Sleep Disorder are', sleep_data['Sleep Disorder'].unique())

The unique values of Sleep Disorder are ['None' 'Sleep Apnea' 'Insomnia']


In [380]:
Sleep_Disorder_count=sleep_data['Sleep Disorder'].value_counts().reset_index()
Sleep_Disorder_count

Sleep Disorder  count
0           None    219
1    Sleep Apnea     78
2       Insomnia     77

In [381]:
gender_disorder = sleep_data.groupby(['Gender', 'Sleep Disorder']).size().reset_index(name='count')
gender_disorder

Gender Sleep Disorder  count
0  Female       Insomnia     36
1  Female           None     82
2  Female    Sleep Apnea     67
3    Male       Insomnia     41
4    Male           None    137
5    Male    Sleep Apnea     11

#### Insights💡
#### 🔸 Insomnia: Both genders experience insomnia, with slightly more cases reported among males (41 instances) compared to females (36 instances).
#### 🔸 No Sleep Disorder: Males report a significantly higher number of instances with no reported sleep disorder (137 instances) compared to females (82 instances).
#### 🔸 Sleep Apnea: Females show a substantially higher prevalence of sleep apnea, with 67 instances reported, compared to only 11 instances among males.

In [382]:
position_disorder = sleep_data.groupby(['Sleep Disorder', 'Occupation']).size().unstack(fill_value=0)
position_disorder

Occupation      Accountant  Doctor  Engineer  Lawyer  Manager  Nurse  \
Sleep Disorder                                                         
Insomnia                 7       3         5       2        0      3   
None                    30      64        57      42        1      9   
Sleep Apnea              0       4         1       3        0     61   

Occupation      Sales Representative  Salesperson  Scientist  \
Sleep Disorder                                                 
Insomnia                           0           29          0   
None                               0            2          2   
Sleep Apnea                        2            1          2   

Occupation      Software Engineer  Teacher  
Sleep Disorder                              
Insomnia                        1       27  
None                            3        9  
Sleep Apnea                     0        4

#### Insights💡
#### 🔸 Insomnia: Notable instances reported among salespersons (29) and teachers (27).
#### 🔸 No Sleep Disorder: Predominantly reported by professionals like doctors (64) and engineers (57), with relatively fewer instances among other occupations.
#### 🔸 Sleep Apnea: Nurses exhibit the highest prevalence (61), followed by doctors (4) and lawyers (3), suggesting potential occupational health concerns within these professions.